e2e/nomostest/portforwarder/port_forwarder.go

```diff
@@ -190,67 +214,165 @@ func (pf *PortForwarder) portForwardToPod(pod string, async bool) error {
-				// tries to use the local port.
+				eventChan <- Event{Type: ErrorEvent, Message: err.Error()}
 				return
 			}
+			if err != nil {
+				pf.logger.Infof("port-forward must be restarted for Pod %s/%s: %v", pf.ns, pf.pod, err)
+			} else {
+				pf.logger.Infof("port-forward must be restarted for Pod %s/%s: nil error", pf.ns, pf.pod)
+			}
 		}
 	}()
-	return nil
+	return eventChan
+}
+
+// startAndWait will start the port-forward and block until seeing an indicator
+// that the port-forward is no longer healthy. If that occurs, it will return.
+func (pf *PortForwarder) startAndWait(eventChan chan Event) error {
+	defer pf.notReady()
+	if err := pf.portForwardToDeployment(); err != nil {
+		return err
+	}
+	eventChan <- Event{
+		Type:    InitEvent,
+		Message: fmt.Sprintf("PortForwarder listening on localhost:%d", pf.localPort),
+	}
+	return pf.waitForFailure()
+}
+
+// waitForFailure waits for any failure condition which indicates the port-forward
+// must be restarted. The port-forward must be restarted if:
+// - the subprocess is killed
+// - the pod is restarted
+func (pf *PortForwarder) waitForFailure() (err error) {
+	ctx, cancel := context.WithCancel(pf.ctx)
+	cmdChan := wrapWithChannel(func() error {
+		return pf.waitForCmdFailure(ctx)
+	})
+	podChan := wrapWithChannel(func() error {
+		return pf.waitForPodFailure(ctx)
+	})
+	defer func() {
+		// before we return, cancel and wait for goroutines to exit
+		cancel()
+		// If any channel returns a TerminalError, return it instead of original error.
+		// This informs upstream logic that a fatal error has occurred, and we must exit the retry loop.
+		cmdErr, ok := <-cmdChan
+		if ok && isTerminalError(cmdErr) {
+			err = cmdErr
+		}
+		podErr, ok := <-podChan
+		if ok && isTerminalError(podErr) {
+			err = podErr
+		}
+	}()
+	// wait for any of the channels to return. a happy little race
+	select {
+	case err = <-cmdChan:
+		return err
+	case err = <-podChan:
+		return err
+	}
+}
+
+// waitForCmdFailure waits for the port-forward subprocess to be killed until ctx is closed.
+// If ctx is closed, the subprocess will be killed before returning.
+// An error is returned to trigger port-forward restart.
+func (pf *PortForwarder) waitForCmdFailure(ctx context.Context) error {
+	// cmd.Wait does not accept a context, so create a wrapper that cancels with the ctx
+	cmdWaitChan := wrapWithChannel(pf.cmd.Wait)
+	var chanErr error
+	select {
+	case <-ctx.Done():
+		// stop waiting for the command to exit, so we can kill it
+		chanErr = ctx.Err()
+	case err := <-cmdWaitChan:
+		return err
+	}
+	// kill the subprocess. This should force the cmd.Wait goroutine to return, if it hasn't already.
+	if err := pf.cmd.Process.Kill(); err != nil && !errors.Is(err, os.ErrProcessDone) {
+		return newTerminalError("failed to kill port-forward process: %v", err)
+	} else if err == nil {
+		pf.logger.Infof("stopped port-forward process for %s/%s", pf.ns, pf.pod)
```

Comment:
Move this up above the Kill call and log it as `killing the port-forward process for %s/%s`.
This could be debug, but I think for now it's ok as Info, since we'll need to debug in CI if it fails.

Reply:
done